core/src/main/java/ai/timefold/solver/core/impl/move/streams/dataset/ForEachDataStream.java

```diff
@@ -12,48 +11,40 @@
 import ai.timefold.solver.core.impl.bavet.uni.ForEachExcludingUnassignedUniNode;
 import ai.timefold.solver.core.impl.bavet.uni.ForEachFromSolutionUniNode;
 import ai.timefold.solver.core.impl.bavet.uni.ForEachIncludingUnassignedUniNode;
-import ai.timefold.solver.core.impl.bavet.uni.ForEachStaticUniNode;
+import ai.timefold.solver.core.impl.move.streams.FromSolutionValueCollectingFunction;
 import ai.timefold.solver.core.impl.move.streams.dataset.common.DataNodeBuildHelper;
-import ai.timefold.solver.core.impl.move.streams.maybeapi.stream.SolutionExtractor;
 
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+@NullMarked
 public final class ForEachDataStream<Solution_, A>
         extends AbstractUniDataStream<Solution_, A>
         implements TupleSource {
 
     private final Class<A> forEachClass;
-    private final Predicate<A> filter;
-    private final SolutionExtractor<Solution_, A> extractor;
-    private final Collection<A> source;
+    private final @Nullable FromSolutionValueCollectingFunction<Solution_, A> valueCollectingFunction;
```

Comment:
Why not to add specialized classes, one for the predicate and another for the FromSolutionValueCollectingFunction?

Reply:
Consistency with the rest of the streams code. (CS also does it like that.) I guess the original idea was to avoid the boilerplate of dozens of different stream classes, but I don't really remember anymore.